### PR TITLE
feat: add `useInfiniteList` hook

### PR DESCRIPTION
**Pull request by yildirayunlu**
The `useInfinityGetList` hook can create lists that can "load more data" or "infinitely scroll".

- [x] Tests
- [x] Docs

### CODE REVIEW

File: packages/core/src/hooks/data/useInfinityGetList.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                });
+            },
+            getNextPageParam: (lastPage) => {
+                const { pagination } = lastPage;
+                const current = pagination?.current || 1;
+                const pageSize = pagination?.pageSize || 10;
+
+                const totalPages = Math.ceil((lastPage.total || 0) / pageSize);
+                return current < totalPages ? Number(current) + 1 : undefined;
+            },
```
Comment (salihozdemir) [inline packages/core/src/hooks/data/useInfinityGetList.ts:203]: If user doesn't return the `pagination` params on their `getList` method. I think, it should use `total` param to calculate `totalPages`.

File: packages/core/src/hooks/data/useInfinityGetList.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            getPreviousPageParam: (lastPage) => {
+                const { pagination } = lastPage;
+                const current = pagination?.current || 1;
+
+                return current === 1 ? undefined : current - 1;
+            },
```
Comment (salihozdemir) [inline packages/core/src/hooks/data/useInfinityGetList.ts:209]: Likewise, if the `pagination` value doesn't return, it will not work.

**Review by salihozdemir — APPROVED** [review]

File: packages/core/src/hooks/data/useInfinityGetList.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * It uses the `getList` method as the query function from the `dataProvider` which is passed to `<Refine>`.
+ *
+ * @see {@link https://refine.dev/docs/core/hooks/data/useInfinityGetList} for more details.
+ *
+ * @typeParam TData - Result data of the query extends {@link https://refine.dev/docs/core/interfaceReferences#baserecord `BaseRecord`}
+ * @typeParam TError - Custom error object that extends {@link https://refine.dev/docs/core/interfaceReferences#httperror `HttpError`}
+ *
+ */
+export const useInfinityGetList = <
```
Comment (omeraplak) [inline packages/core/src/hooks/data/useInfinityGetList.ts:73]:
```suggestion
export const useInfinityList = <
```

File: examples/infinity-list/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+[![Twitter Follow](https://img.shields.io/twitter/follow/refine_dev?style=social)](https://twitter.com/refine_dev)
+
+<a href="https://www.producthunt.com/posts/refine-3?utm_source=badge-top-post-badge&utm_medium=badge&utm_souce=badge-refine&#0045;3" target="_blank"><img src="https://api.producthunt.com/widgets/embed-image/v1/top-post-badge.svg?post_id=362220&theme=light&period=daily" alt="refine - 100&#0037;&#0032;open&#0032;source&#0032;React&#0032;framework&#0032;to&#0032;build&#0032;web&#0032;apps&#0032;3x&#0032;faster | Product Hunt" style="width: 250px; height: 54px;" width="250" height="54" /></a>
+
+</div>
+
+## Try it out on your local
+
+```bash
+npm create refine-app@latest -- --example base-headless
```
Comment (omeraplak) [inline examples/infinity-list/README.md:37]:
```suggestion
npm create refine-app@latest -- --example base-headless
```
we need to update

File: examples/infinity-list/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+## Try it out on StackBlitz
+
+<br/>
+
+[![Open base-headless example from refine](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/refinedev/refine/tree/master/examples/base-headless?terminal=start&preset=node)
```
Comment (omeraplak) [inline examples/infinity-list/README.md:44]: we need to update

File: examples/infinity-list/public/index.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  <!--
+      Notice the use of %PUBLIC_URL% in the tags above.
+      It will be replaced with the URL of the `public` folder during the build.
+      Only files inside the `public` folder can be referenced from the HTML.
+
+      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+  <title>refine base headless example</title>
```
Comment (omeraplak) [inline examples/infinity-list/public/index.html:25]: we need to update

File: examples/infinity-list/public/manifest.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "refine base infinity list example",
+  "name": "refine base infinity list example",
```
Comment (omeraplak) [inline examples/infinity-list/public/manifest.json:3]: we need to update

File: packages/core/src/hooks/data/useInfinityGetList.ts [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                    description: err.message,
+                    type: "error",
+                });
+            },
+            getNextPageParam: (lastPage) => {
+                const { pagination } = lastPage;
+                const current = pagination?.current || 1;
+                const pageSize = pagination?.pageSize || 10;
+
+                const totalPages = Math.ceil((lastPage.total || 0) / pageSize);
```
Comment (omeraplak) [inline packages/core/src/hooks/data/useInfinityGetList.ts:201]: we can move this calculation to a helper

**Review by omeraplak — APPROVED** [review]

File: packages/core/src/contexts/data/IDataContext.ts [hunk 9]
```diff
@@ -87,6 +87,15 @@ export interface CustomResponse<TData = BaseRecord> {
 export interface GetListResponse<TData = BaseRecord> {
     data: TData[];
     total: number;
+    pageInfo?: {
+        hasNextPage: boolean;
+        hasPreviousPage: boolean;
+    };
+}
+
+export interface GetInfinityListResponse<TData = BaseRecord>
```
Comment (aliemir) [inline packages/core/src/contexts/data/IDataContext.ts:96]:
```suggestion
export interface GetInfiniteListResponse<TData = BaseRecord>
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+} from "@definitions/helpers";
+
+export interface UseInfinityConfig {
+    pagination?: Pagination;
+    hasPagination?: boolean;
+    sort?: CrudSorting;
+    filters?: CrudFilters;
+}
+
+export type UseInfinityListProps<TData, TError> = {
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts:39]:
```suggestion
export type UseInfiniteListProps<TData, TError> = {
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    useDataProvider,
+} from "@hooks";
+import {
+    queryKeys,
+    pickDataProvider,
+    getNextPageParam,
+    getPreviousPageParam,
+} from "@definitions/helpers";
+
+export interface UseInfinityConfig {
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts:32]:
```suggestion
export interface UseInfiniteListConfig {
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+export type UseInfinityListProps<TData, TError> = {
+    /**
+     * Resource name for API data interactions
+     */
+    resource: string;
+    /**
+     * Configuration for pagination, sorting and filtering
+     * @type [`UseInfinityConfig`](/docs/api-reference/core/hooks/data/useInfiniteList/#config-parameters)
+     */
+    config?: UseInfinityConfig;
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts:48]:
```suggestion
    config?: UseInfiniteListConfig;
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    /**
+     * react-query's [useInfiniteQuery](https://tanstack.com/query/v4/docs/react/reference/useInfiniteQuery) options,
+     */
+    queryOptions?: UseInfiniteQueryOptions<
+        GetInfinityListResponse<TData>,
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts]:
```suggestion
        GetInfiniteListResponse<TData>,
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+import {
+    CrudFilters,
+    Pagination,
+    BaseRecord,
+    HttpError,
+    CrudSorting,
+    MetaDataQuery,
+    SuccessErrorNotification,
+    LiveModeProps,
+    GetInfinityListResponse,
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts]:
```suggestion
    GetInfiniteListResponse,
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    config,
+    queryOptions,
+    successNotification,
+    errorNotification,
+    metaData,
+    liveMode,
+    onLiveEvent,
+    liveParams,
+    dataProviderName,
+}: UseInfinityListProps<TData, TError>): InfiniteQueryObserverResult<
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts:92]:
```suggestion
}: UseInfiniteListProps<TData, TError>): InfiniteQueryObserverResult<
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+    GetInfinityListResponse<TData>,
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts]:
```suggestion
    GetInfiniteListResponse<TData>,
```

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+            ...liveParams,
+        },
+        channel: `resources/${resource}`,
+        enabled: isEnabled,
+        liveMode,
+        onLiveEvent,
+    });
+
+    const queryResponse = useInfiniteQuery<
+        GetInfinityListResponse<TData>,
```
Comment (aliemir) [inline packages/core/src/hooks/data/useInfiniteList.ts]:
```suggestion
        GetInfiniteListResponse<TData>,
```

**Review by aliemir — APPROVED** [review]
❤️

**Review by alicanerdurmaz — APPROVED** [review]

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+export type UseInfiniteListProps<TData, TError> = {
+    /**
+     * Resource name for API data interactions
+     */
+    resource: string;
+    /**
+     * Configuration for pagination, sorting and filtering
+     * @type [`useInfiniteListConfig`](/docs/api-reference/core/hooks/data/useInfiniteList/#config-parameters)
+     */
+    config?: UseInfiniteListConfig;
```
Comment (omeraplak) [inline packages/core/src/hooks/data/useInfiniteList.ts:48]: Should we change the config property? @aliemir @salihozdemir

Comment (salihozdemir) [reply packages/core/src/hooks/data/useInfiniteList.ts:48]: In my opinion, we should change.

File: packages/core/src/definitions/helpers/useInfinitePagination/index.ts [hunk 19]
```diff
@@ -1,26 +1,30 @@
...
-export const getPreviousPageParam = (lastPage: GetInfiniteListResponse) => {
-    const { pagination, pageInfo } = lastPage;
-    const current = pagination?.current || 1;
+export const getPreviousPageParam = (lastPage: GetListResponse) => {
+    const { pagination, cursor } = lastPage;
 
-    if (pageInfo) {
-        return pageInfo.hasPreviousPage ? current - 1 : undefined;
+    // cursor pagination
+    if (cursor && cursor.prev) {
```
Comment (omeraplak) [inline packages/core/src/definitions/helpers/useInfinitePagination/index.ts:23]:
```suggestion
    if (cursor?.prev) {
```

File: examples/use-infinite-list/src/github-data-provider/index.ts [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+export const githubDataProvider = (
+    httpClient: AxiosInstance = axiosInstance,
+): Omit<
+    Required<DataProvider>,
+    "createMany" | "updateMany" | "deleteMany"
+> => ({
+    getList: async ({ resource, pagination }) => {
+        const { data } = await httpClient.get(
+            `https://api.github.com/repos/refinedev/refine/${resource}?until=${
```
Comment (omeraplak) [inline examples/use-infinite-list/src/github-data-provider/index.ts:17]:
```suggestion
            `https://api.github.com/${resource}?until=${
```

File: examples/use-infinite-list/src/pages/commits/list.tsx [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+export const CommitList: React.FC = () => {
+    const {
+        data,
+        error,
+        hasNextPage,
+        isLoading,
+        fetchNextPage,
+        isFetchingNextPage,
+    } = useInfiniteList({
+        resource: "commits",
```
Comment (omeraplak) [inline examples/use-infinite-list/src/pages/commits/list.tsx:13]:
```suggestion
        resource: "repos/refinedev/refine/commits",
```

File: examples/use-infinite-list/src/pages/commits/list.tsx [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+                return lastCommit.commit.committer.date;
+            },
+        },
+    });
+
+    if (isLoading) {
+        return <p>Loading</p>;
+    }
+    if (error) {
+        return <p>ERROR</p>;
```
Comment (omeraplak) [inline examples/use-infinite-list/src/pages/commits/list.tsx:28]:
```suggestion
        return <p>Something went wrong</p>;
```

File: documentation/docs/api-reference/core/hooks/data/useInfiniteList.md [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+| `contains`   | Contains                        |
+| `ncontains`  | Doesn't contain                 |
+| `containss`  | Contains, case sensitive        |
+| `ncontainss` | Doesn't contain, case sensitive |
+| `null`       | Is null or not null             |
+
+<br />
+
+:::tip
+`useInfiniteList` can also accept all `useInfiniteQuery` options as a third parameter.
```
Comment (salihozdemir) [inline documentation/docs/api-reference/core/hooks/data/useInfiniteList.md:295]:
```suggestion
`useInfiniteList` can also accept all `useInfiniteQuery` options as a parameter.  
```

File: documentation/docs/api-reference/core/hooks/data/useInfiniteList.md [hunk 24]
```diff
@@ -0,0 +1,375 @@
...
+        },
+    ],
+});
+
+render(<RefineHeadlessDemo />);
+```
+
+## Usage
+
+Let's say that we have a resource named `posts`
```
Comment (salihozdemir) [inline documentation/docs/api-reference/core/hooks/data/useInfiniteList.md:87]:
```suggestion
Let's assume that we have a `posts` resource with the following data:
```

File: documentation/docs/api-reference/core/hooks/data/useInfiniteList.md [hunk 25]
```diff
@@ -0,0 +1,375 @@
...
+        ]
+    },
+    "hasNextPage": true,
+    "hasPreviousPage": false,
+    "isFetchingNextPage": false,
+    "isFetchingPreviousPage": false
+    ...
+}
+```
+
```
Comment (salihozdemir) [inline documentation/docs/api-reference/core/hooks/data/useInfiniteList.md:154]: If your API returns the result like above, you can use `useInfiniteList` without any configuration.

Maybe we can say like this.

**Review by aliemir — APPROVED** [review]

File: packages/core/src/hooks/data/useInfiniteList.ts [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+
+                const notificationConfig =
+                    typeof errorNotification === "function"
+                        ? errorNotification(err, { metaData, config }, resource)
+                        : errorNotification;
+
+                handleNotification(notificationConfig, {
+                    key: `${resource}-useInfiniteList-notification`,
+                    message: translate(
+                        "common:notifications.error",
```
Comment (sevkioruc) [inline packages/core/src/hooks/data/useInfiniteList.ts:190]: Hey @yildirayunlu , @omeraplak 
Should it been changed as `notifications.error` according to this issue(https://github.com/refinedev/refine/issues/3328)?

Comment (omeraplak) [reply packages/core/src/hooks/data/useInfiniteList.ts:190]: Yes, you are right 👍

Comment (sevkioruc) [reply packages/core/src/hooks/data/useInfiniteList.ts:190]: is need to open an issue before fix? @omeraplak

Comment (yildirayunlu) [reply packages/core/src/hooks/data/useInfiniteList.ts:190]: Thank you for feedback @sevkioruc. Fixed this [PR](https://github.com/refinedev/refine/pull/3455).

**Pull request merged**